src/Datadog.Trace.ClrProfiler.Managed/CallTarget/Handlers/IntegrationMapper.cs

```diff
@@ -0,0 +1,714 @@
+            }
+
+            ParameterInfo[] onMethodBeginParameters = onMethodBeginMethodInfo.GetParameters();
+            if (onMethodBeginParameters.Length < argumentsTypes.Length)
+            {
+                throw new ArgumentException($"The method: {BeginMethodName} with {onMethodBeginParameters.Length} paremeters in type: {integrationType.FullName} has less parameters than required.");
+            }
+            else if (onMethodBeginParameters.Length > argumentsTypes.Length + 1)
+            {
+                throw new ArgumentException($"The method: {BeginMethodName} with {onMethodBeginParameters.Length} paremeters in type: {integrationType.FullName} has more parameters than required.");
+            }
+            else if (onMethodBeginParameters.Length != argumentsTypes.Length && onMethodBeginParameters[0].ParameterType != genericArgumentsTypes[0])
+            {
+                throw new ArgumentException($"The first generic argument for method: {BeginMethodName} in type: {integrationType.FullName} must be the same as the first parameter for the instance value.");
+            }
+
+            List<Type> callGenericTypes = new List<Type>();
+
+            bool mustLoadInstance = onMethodBeginParameters.Length != argumentsTypes.Length;
+            Type instanceGenericType = genericArgumentsTypes[0];
+            Type instanceGenericConstraint = instanceGenericType.GetGenericParameterConstraints().FirstOrDefault();
+            Type instanceProxyType = null;
+            if (instanceGenericConstraint != null)
+            {
+                var result = DuckType.GetOrCreateProxyType(instanceGenericConstraint, targetType);
+                instanceProxyType = result.ProxyType;
+                callGenericTypes.Add(instanceProxyType);
+            }
+            else
+            {
+                callGenericTypes.Add(targetType);
+            }
+
+            DynamicMethod callMethod = new DynamicMethod(
+                     $"{onMethodBeginMethodInfo.DeclaringType.Name}.{onMethodBeginMethodInfo.Name}",
+                     typeof(CallTargetState),
+                     new Type[] { targetType }.Concat(argumentsTypes),
+                     onMethodBeginMethodInfo.Module,
+                     true);
+
+            ILGenerator ilWriter = callMethod.GetILGenerator();
+
+            // Load the instance if is needed
+            if (mustLoadInstance)
+            {
+                ilWriter.Emit(OpCodes.Ldarg_0);
+
+                if (instanceGenericConstraint != null)
+                {
+                    ConstructorInfo ctor = instanceProxyType.GetConstructors()[0];
+                    if (targetType.IsValueType && !ctor.GetParameters()[0].ParameterType.IsValueType)
+                    {
+                        ilWriter.Emit(OpCodes.Box, targetType);
+                    }
+
+                    ilWriter.Emit(OpCodes.Newobj, ctor);
```

Comment:
Since this pattern shows up in the for loop as well, perhaps this can be refactored into an IL helper method?

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/1007/commits/6cf2259ec176e57423a1837f3a0b282eb73361a4